xsnap/sources/xsnap-worker.c

```diff
@@ -147,6 +147,67 @@ typedef enum {
 	E_TOO_MUCH_COMPUTATION = 17,
 } ExitCode;
 
+#define MAX_TIMESTAMPS 100
+static struct timeval timestamps[MAX_TIMESTAMPS];
+static unsigned int num_timestamps;
+static unsigned int timestamps_overrun;
+static void resetTimestamps() {
+	timestamps_overrun = 0;
+	num_timestamps = 0;
+	// on 64-bit platforms, 'struct timeval' usually needs 8+8=16 bytes
+	//printf("sizeof(time_t): %ld\n", sizeof(time_t));
+	//printf("sizeof(time_suseconts_t): %ld\n", sizeof(suseconds_t));
+}
+static void recordTimestamp() {
+	if (num_timestamps < MAX_TIMESTAMPS) {
+		gettimeofday(&(timestamps[num_timestamps]), NULL);
```

Comment:
What is the resolution of this syscall? Is it monotonic? Which Node API do we use on the other side, and do we know which syscall that corresponds to?

Reply:
According to `man gettimeofday`, it returns a `struct tv`, which is a `time_t` of seconds and a `suseconds_t` of microseconds. Experimentally I observe both `sizeof time_t` and `sizeof suseconds_t` reporting 8 bytes, so a 64bit integer each (on current macOS, M1Pro CPU). The resolution is platform dependent, but I think everything modern provides actual microseconds. There are syscalls to do better, but this one is super-standard/portable, and I'm not currently concerned about anything smaller than microseconds (although I'll be delighted if we get everything else running fast enough to make us care about nanoseconds).

In our application (swingset), the Node API on the other side uses `perf_hooks.performance`, and adds the constant `performance.timeOrigin` to the current value of `performance.now()`.

I spent the afternoon tracing down how these are managed within Node (and I'm reminded that Time is one of the three deadly problems in computer science, which seems like it should be trivial, but in fact the complexity wants to kill you.. strings and money are the other two on my list).

Node bottoms out into two POSIX functions. The first is the same `gettimeofday()` I used in xsnap.c . The second is `clock_gettime(CLOCK_MONOTONIC)`, which POSIX specifies to not decrease, but is still subject to frequency adjustments. Experimentally, CLOCK_MONOTONIC appears to return a structure with seconds/nanoseconds since boot.

At startup, every Node.js process samples `clock_gettime(CLOCK_MONOTONIC)` followed immediately by `gettimeofday()`: the two calls are right next to each other in `src/node_perf.cc`, but there's still probably a delay of a few ns between them (call this the "sampling delta"). The CLOCK_MONOTONIC value is storted in a variable named `timeOrigin`, and is used as the offset for calls to `performance.now()`: every time you call `now()`, it returns `clock_gettime(CLOCK_MONOTONIC) - timeOrigin`, so you get time-since-process-start. This protects `now()` from timequakes (when something like NTP changes the host's clock, either slowly or drastically).

By adding that value to `performance.timeOrigin`, we get time-since-epoch, plus the sampling delta. I think this is the recommended way to get high-resolution time-since-epoch in Node. The `process.hrtime.bigint()` returns the CLOCK_MONOTONIC value directly, which is not since-epoch. `Date.now()` has only millisecond resolution. But the sum will drift from a fresh sampling of `gettimeofday()` with each timequake, and won't line up again until the process is restarted (or you use 

So we're comparing repeated calls to `gettimeofday()` (on the worker side), against a single `gettimeofday()` plus a delta of `CLOCK_MONOTONIC` (on the kernel side). One systemic offset is the sampling delta. Another potential offset is if/when `gettimeofday()` has a timequake (something like NTP changes the host's clock). Those offsets may get larger and larger over time, until the kernel is restarted and gets a fresh sample of `gettimeofday()` (there's a `performance.refreshTimeOrigin()`, but it doesn't cause a new call to `gettimeofday()`, so I think it's broken/irrelevant).

If I care enough to avoid that, I think we have two options:

* 1: change the `xsnap.c` side to perform the same double-sampling at startup, and build seconds-since-epoch by addition just like Node.js does.
  * This would be additionally vulnerable to drift between the two processes when either one restarts (and does a new `gettimeofday()`).
  * On the plus side, the API would continue to be in seconds-since-epoch, which is the most meaningful thing to report
* 2: change the interface to report CLOCK_MONOTONIC values in the metering data, ignoring the epoch, and having the kernel side do the addition so it can record an absolute time in the slogfile
  * it's safe to compare `CLOCK_MONOTONIC` against `CLOCK_MONOTONIC`, but the Node.js API doesn't actually expose that: only deltas between `CLOCK_MONOTONIC` values
  * the worker process could do its own subtraction, to build a value that behaves like `performance.now()`, but that'd be worse: you can't compare seconds-since-process-1-start against seconds-since-process-2-start

So really, the question is whether to report each `gettimeofday()`, or to add a single `gettimeofday()` to a delta between `CLOCK_MONOTONIC` values.

FWIW, my current tests aren't showing me any surprises: the timestamps I get from the worker line up in the middle of the timestamps I'm seeing on the kernel process. But I'm only looking at the timestamps from a single unit test. I should find a way to modify the system time in the middle of a test and see how bad it can get.

Reply:
I'm going to do the simple and most-stable thing on the `xsnap` side: just report `gettimeofday()`. I'll investigate how to match this correctly on the swingset side as I make the corresponding swingset changes.

Reply:
I read a bit more about `CLOCK_MONOTONIC`: linux implements it by returning the normal `gettimeofday()` time plus an offset, which is initialized to zero at kernel boot. If someone calls `settimeofday()`, the kernel calculates the delta being applied to the main clock, and reduces the offset by that delta, so the net `CLOCK_MONOTONIC` value doesn't change. NTP generally only jumps the time once, when you first install it, and it finds that it needs to make a change of more than 500ms. That jump will get included in the offset. After that, NTP usually manages to limit itself to gentle slews by adjusting the tick update frequency, and since those don't use `settimeofday()`, the offset isn't changed. So unless you're unlucky enough to observe a leap second or a time source flapping, or you're on a laptop that sleeps, or you've got a time synchronization tool that uses `settimeofday()` a lot instead of `adjtime()`, then `gettimeofday` and `timeOrigin + CLOCK_MONOTONIC` should remain within a tiny offset of each other.

Still, I found an NPM module (`microtime`) to use `gettimeofday()` on the swingset side, and I'll change the slogger to use it. I'm more interested in maintaining the comparability of slog timestamps to other processes and computers than I am worried about the occasional timequake causing an outlier in a few spans. We do lots of analysis that subtracts e.g. `deliver-result` - `deliver`, or `syscall-result` - `syscall`, and if a timequake occurs during a delivery or syscall, that one value will be corrupted. By using `gettimeofday()` on both sides, we'll limit the damage to a single delivery, whereas if we used `timeOrigin + CLOCK_MONOTONIC`, we'd see the offset present in every single timestamp until the process restarted.